kafka.go

```diff
@@ -1,161 +1,213 @@
+		EventName: publishBuilder.eventName,
+		Namespace: publishBuilder.namespace,
+		ClientID:  publishBuilder.clientID,
+		UserID:    publishBuilder.userID,
+		Timestamp: time.Now(),
+		Version:   version,
+		Payload:   publishBuilder.payload,
 	}
-}
 
-// Length returns the length of the event entry
-func (evtEntry *eventEntry) Length() int {
-	evtEntry.ensureEncoded()
-	return len(evtEntry.encoded)
-}
+	eventBytes, err := marshal(event)
+	if err != nil {
+		return kafka.Message{}, err
+	}
+
+	return kafka.Message{
+		Key:   []byte(id),
+		Value: eventBytes,
+	}, nil
 
-// Encode serialize the struct
-func (evtEntry *eventEntry) Encode() ([]byte, error) {
-	evtEntry.ensureEncoded()
-	return evtEntry.encoded, evtEntry.err
 }
 
-// PublishEvent push an event to a Kafka topic synchronously
-func (client *KafkaClient) PublishEvent(event *Event) error {
-	event.Realm = client.realm
-	event.Time = time.Now().UTC()
+// Register register callback and consume a topic
+func (client *KafkaClient) Register(subscribeBuilder *SubscribeBuilder) {
+	topic := constructTopic(client.realm, subscribeBuilder.topic)
+	if client.registerCallback(topic, subscribeBuilder.eventName, subscribeBuilder.callback) {
+		logrus.Warnf("topic and event already registered. event : %s", subscribeBuilder.eventName)
+		return
+	}
 
-	_, _, err := client.syncProducer.SendMessage(&sarama.ProducerMessage{
-		Timestamp: time.Now().UTC(),
-		Value: &eventEntry{
-			Event: *event,
-		},
-		Topic: event.Topic,
-	})
+	config := client.subscribeConfig
+	config.Topic = subscribeBuilder.topic
+	reader := kafka.NewReader(config)
+	if err := reader.SetOffset(kafka.LastOffset); err != nil {
+		logrus.Error("unable to set offset : ", err)
+		return
+	}
 
-	return err
+	// listening a topic
+	// each topic would have his own goroutine
+	go func() {
+		defer func() {
+			_ = reader.Close()
+		}()
+		for {
+			consumerMessage, err := reader.ReadMessage(subscribeBuilder.ctx)
+			if err != nil {
+				logrus.Error("unable to consume topic from kafka : ", err)
+				return
+			}
+			go client.processMessage(consumerMessage)
+		}
+	}()
 }
 
-// PublishEventAsync publish to a Kafka topic asynchronously
-func (client *KafkaClient) PublishEventAsync(event *Event) {
-	event.Realm = client.realm
-	event.Time = time.Now().UTC()
+// registerCallback add callback to map with topic and eventName as a key
+func (client *KafkaClient) registerCallback(topic, eventName string, callback func(event *Event, err error)) (
+	isRegistered bool) {
+	client.Mutex.Lock()
+	if client.subscribeMap == nil {
+		client.subscribeMap = make(map[string]map[string]func(*Event, error))
+	}
 
-	client.asyncProducer.Input() <- &sarama.ProducerMessage{
-		Timestamp: time.Now().UTC(),
-		Value: &eventEntry{
-			Event: *event,
-		},
-		Topic: event.Topic,
+	if _, isExists := client.subscribeMap[topic][eventName]; isExists {
+		client.Mutex.Unlock()
+		return true
 	}
+
+	client.subscribeMap[topic][eventName] = callback
+	client.Mutex.Unlock()
```

Comment:
What do you think if we use defer to unlock the mutex and move it right after you lock? 
If in the future we introduce new return above this, I'm afraid we forget to add unlock.

Reply:
sure, it would be better
thanks